File: app/javascript/channels/prompt_channel.js
import consumer from "./consumer"

let subscriptions = {}

export function subscribeToExecution(executionId, handlers) {
  // Unsubscribe from existing subscription if any
  if (subscriptions[executionId]) {
    subscriptions[executionId].unsubscribe()
  }
  
  subscriptions[executionId] = consumer.subscriptions.create(
    { 
      channel: "PromptChannel",
      execution_id: executionId 
    },
    {
      connected() {
        console.log(`Connected to prompt channel for execution ${executionId}`)
        if (handlers.onConnected) handlers.onConnected()
      },

      disconnected() {
        console.log(`Disconnected from prompt channel for execution ${executionId}`)
        if (handlers.onDisconnected) handlers.onDisconnected()
      },

      received(data) {
        console.log("Received data:", data)
        
        switch(data.type) {
          case 'chunk':
            if (handlers.onChunk) {
              handlers.onChunk(data.iteration, data.content)
            }
            break
            
          case 'complete':
            if (handlers.onComplete) {
              handlers.onComplete(data.iteration, data.result)
            }
            break
            
          case 'error':
            if (handlers.onError) {
              handlers.onError(data.iteration, data.error)
            }
            break
            
          default:
            console.log("Unknown message type:", data.type)
        }
      }
    }
  )
  
  return subscriptions[executionId]
}

export function unsubscribeFromExecution(executionId) {
  if (subscriptions[executionId]) {
    subscriptions[executionId].unsubscribe()
    delete subscriptions[executionId]
  }
}

File: app/channels/prompt_channel.rb
class PromptChannel < ApplicationCable::Channel
  def subscribed
    execution = Execution.find(params[:execution_id])
    stream_for execution
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end
  
  def self.broadcast_update(execution, data)
    broadcast_to(execution, data)
  end
  
  def self.broadcast_chunk(execution, iteration_number, chunk)
    broadcast_to(execution, {
      type: 'chunk',
      iteration: iteration_number,
      content: chunk
    })
  end
  
  def self.broadcast_complete(execution, iteration_number, result)
    broadcast_to(execution, {
      type: 'complete',
      iteration: iteration_number,
      result: result.as_json(except: [:created_at, :updated_at])
    })
  end
  
  def self.broadcast_error(execution, iteration_number, error)
    broadcast_to(execution, {
      type: 'error',
      iteration: iteration_number,
      error: error
    })
  end
end